feat: Allow automatic project actions on events

lunny opened the pull request:
This PR introduces project workflows, allowing users to configure event-based automations directly from the UI.
Users can define filters and actions for each workflow event, and duplicate existing events when needed.

### Concepts

Every workflow is composed of `event`, `filters`, and `actions`.

- **Event** - An Event defines when a workflow should be triggered. It represents specific activities such as an issue being opened, a pull request being merged, or a project column being changed.
- **Filter** - A Filter refines when an event should take effect. It allows users to specify conditions like issue type, labels, or project columns to ensure actions only run in targeted scenarios.
- **Action** - An Action defines what happens when the event and filters match. It performs operations such as adding labels, moving issues between columns, or changing the issue or pull request state.

### Available Events

Currently, the following nine events are supported:

- On issues/pull requests opened
- On issues/pull requests closed
- On issues/pull requests reopened
- On issues/pull requests was added to the project
- On issues/pull requests was removed to the project
- On issues/pull requests' column was changed
- On code changes requested
- On code review approved
- On pull request merged

Each event type supports its own set of filters and actions.

### Available Filters

There are four types of filters that can be configured for an event:

- **Apply to** - Specify the target type: `issues only`, `pull requests only`, or `issues and pull requests`.
- **Labels** - Trigger the event only if the issue or pull request contains one or more specified labels.
- **Move from column** - Filter events based on the original project column.
- **Move to column** - Filter events based on the destination project column.

### Available Actions

There are four types of actions that can be executed when an event is triggered:

- **Add labels** - Automatically add one or more labels to the issue or pull request.
- **Remove labels** - Remove one or more specified labels.
- **Move to column** - Move the issue or pull request to a designated project column.
- **Issue state** - Change the state by closing or reopening the issue or pull request.

### Enable/Disable an event workflow

Each event workflow can be individually enabled or disabled. When disabled, the workflow will be ignored even if the event occurs. This feature is useful when you want to temporarily stop certain automated actions without deleting the workflow configuration.

### Clone an event workflow

When you need multiple workflows for the same event, you can clone an existing event to create a new one with different filters and actions. This allows you to customize how the same event behaves under different conditions, making workflows more flexible and powerful.

### Screen shots

- View mode

<img width="1302" height="656" alt="image" src="https://github.com/user-attachments/assets/2e0caff2-4dbc-4e84-8765-790251858734" />

- Edit mode

<img width="1295" height="818" alt="image" src="https://github.com/user-attachments/assets/35cc130c-372d-46d0-90a3-28b85c95b5a2" />

- Dark theme

<img width="1308" height="814" alt="image" src="https://github.com/user-attachments/assets/dc319bea-ff3c-47e6-b109-dfcdfff91f89" />

- Issue comments triggered by project workflow

<img width="781" height="295" alt="image" src="https://github.com/user-attachments/assets/54f8301b-726d-4bd2-9622-b436386d75cc" />

Additional features:
- Project readers could view the workflows page but only project writer could update them
- Project Workflow APIs

* Resolve #14359
* Resolve #26704
* Resolve #27990
* Replace  #28745

Documentation PR https://gitea.com/gitea/docs/pulls/308

lunny commented:
I'm now hesitate to add this feature because looks like some actions can do the same things. Like https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/automating-projects-using-actions and more third-party actions can do it easier.

rohrschacht commented:
I would really like to see this merged, I am waiting for the ability to automatically set a project for a new issue via the issue template (#25028). GitHub provides this through an "org-name/BOARD_ID" syntax as announced [here](https://github.blog/changelog/2023-08-10-github-issues-projects-august-10th-update/#updates-to-issue-forms). I think having to set up an action for that (which would also require to set up a runner?) is quite cumbersome for such a simple feature.

silkentrance commented:
> I'm now hesitate to add this feature because looks like some actions can do the same things. Like https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/automating-projects-using-actions and more third-party actions can do it easier.

I do not know, but firing up multiple containers to just set a specific label or remove a label, or even re-open or close an issue as part of the workflow seems to be overkill to me, what do you think?

thekk1 commented:
> I'm now hesitate to add this feature because looks like some actions can do the same things. Like https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/automating-projects-using-actions and more third-party actions can do it easier.

What's the status here now? This is quite a wanted feature.
Do you have just a feeling about this or a specific good reason?
Running multiple containers for just moving around a Issue is for sure not the best practice.

lunny commented:
> > I'm now hesitate to add this feature because looks like some actions can do the same things. Like https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/automating-projects-using-actions and more third-party actions can do it easier.
> 
> What's the status here now? This is quite a wanted feature. Do you have just a feeling about this or a specific good reason? Running multiple containers for just moving around a Issue is for sure not the best practice.

Yes, this is a good reason to use the internal implementation rather than CI/CD. I will continue this work.

thekk1 commented:
> > > I'm now hesitate to add this feature because looks like some actions can do the same things. Like https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/automating-projects-using-actions and more third-party actions can do it easier.
> > 
> > 
> > What's the status here now? This is quite a wanted feature. Do you have just a feeling about this or a specific good reason? Running multiple containers for just moving around a Issue is for sure not the best practice.
> 
> Yes, this is a good reason to use the internal implementation rather than CI/CD. I will continue this work.

Great, thx for your feedback. I am really looking forward to it.

thekk1 commented:
As far as I understand the description, this will depend on Workflows?
So Actions has to be enabled?
This would be a problem for us, because the workflows in our repos are for the environment of our customers and Actions are always disabled because they will not work in our environment.

lunny commented:
> As far as I understand the description, this will depend on Workflows? So Actions has to be enabled? This would be a problem for us, because the workflows in our repos are for the environment of our customers and Actions are always disabled because they will not work in our environment.

No. This is project workflow which is different from Actions workflow.

lunny commented:
And what's better? Storing the workflows in the special YAML files on the default branch or storing them in the database? This is a block to prevent me from continuing the work.

thekk1 commented:
I would purpose to store them in the database. Branches are for code changes and depending stuff but this is a projekt management feature that is specific for gitea. i.e.: If you mirror your code to an other git system like github the projekt workflows are only garbage.

eeyrjmr commented:
> I would purpose to store them in the database. Branches are for code changes and depending stuff but this is a projekt management feature that is specific for gitea. i.e.: If you mirror your code to an other git system like github the projekt workflows are only garbage there.

but storing in the db will require some form of UI  to create/edit.  
Right now .gitea/* is used to store issue templates and all of that is "garbage" if you were to mirror into github since github only looks in .github thus that objection is moot.

thekk1 commented:
> but storing in the db will require some form of UI to create/edit. Right now .gitea/* is used to store issue templates and all of that is "garbage" if you were to mirror into github since github only looks in .github thus that objection is moot.

For my defence I didn't know about the templates and the storage path. In that case I would agree to that argument.

Woonhyung59 commented:
Thank you for working on this important feature. Many users, including myself, are eagerly anticipating the addition of project workflows to Gitea. This functionality would greatly enhance project management capabilities within the platform.

silkentrance commented:
@lunny is there a wiki page aka spec where the actions etc. are defined, that one can take on a specific workflow event / external event / transition / etc.

lunny commented:
> @lunny is there a wiki page aka spec where the actions etc. are defined, that one can take on a specific workflow event / external event / transition / etc.

Not exactly, but I’ll submit another PR to [Gitea Docs](https://gitea.com/gitea/docs) to update it.

JohnDowson commented:
Is there a timeline for this feature? The lack of automation in the project board is about the only thing holding me off on recommending Gitea for my org as part of out JB Space Replacement Program

lunny commented:
> Is there a timeline for this feature? The lack of automation in the project board is about the only thing holding me off on recommending Gitea for my org as part of our JB Space Replacement Program

It will be expected to be merged before the 1.24 release which should be on 4/30/2025

steveepreston commented:
Please don't ignore this 🙏

msi-tl commented:
After missing 1.24 could you please make sure this important, wanted feature will it make in 1.25?

coldhearti commented:
Bumping this also, our org is actively using the kanban and automations would level up the user experience manyfold!

lunny commented:
> @lunny are you still working on this PR? What is still missing? Do you need any help or support?

I’m still working on this, but it’s not a high priority for me at the moment. I’d appreciate it if you or other contributors could make some improvements after my next push. I’ll be pushing a new update soon.

go-blue-2023 commented:
What would it take to make this one a priority. As others have commented this is a feature that a lot of people really want and are waiting on.

steveepreston commented:
@lunny ❤️

lunny commented:
This is ready to review now.

lunny commented:
@tpokorra Thank you for your testing and feedback. I’ve addressed the first and third issues. I left the second one open because it will require additional work. The page should refresh even when labels are added from the issue UI, so I believe it’s better handled in a separate PR.

wxiaoguang commented on models/user/user_system.go:
```diff
@@ -65,12 +65,46 @@ func (u *User) IsGiteaActions() bool {
 
+const (
+	ProjectWorkflowsUserID    int64 = -3
+	ProjectWorkflowsUserName        = "project-workflows"
```
It seems to be a bad name and is likely to conflict with end user's name.

lunny replied:
All system bot type user have a `[bot]` prefix before the name and it should have no profile page. Then it's different from the user who have the same name.

wxiaoguang commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+	if eventID == 0 {
+		// check if workflow event is valid
+		if !project_model.IsValidWorkflowEvent(form.EventID) {
+			ctx.JSON(http.StatusBadRequest, map[string]any{"error": "EventID is invalid"})
```
Why sometimes `"error": "InvalidEventID",` sometimes `"error": "EventID is invalid"`? Can you use consistency designs?

And there is already JSONError to help to avoid low-level mistakes.

wxiaoguang commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+		Actions       []project_model.WorkflowAction          `json:"actions"`
+		Summary       string                                  `json:"summary"` // Human readable filter description
+		Enabled       bool                                    `json:"enabled"`
+		IsConfigured  bool                                    `json:"isConfigured"` // Whether this workflow is configured/saved
```
Why sometimes `display_name`, sometimes `isConfigured`?

lunny replied:
These are two columns. I don't understand what you mean.

wxiaoguang commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+				"actions":      wf.WorkflowActions,
+				"summary":      workflowSummary,
+				"enabled":      wf.Enabled,
+			},
```
No struct for this? It looks the same as `WorkflowConfig`?

lunny replied:
48e144d015

wxiaoguang commented on templates/repo/projects/workflows.tmpl:
```diff
@@ -0,0 +1,12 @@
+{{template "base/head" .}}
+<div role="main" aria-label="{{.Title}}" class="page-content repository projects view-project">
```
Why `view-project`?

tpokorra commented:
@lunny can I support you in any way to fix the issues that came up in the review?

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+	}
+
+	// Get repository labels
+	labels, err := issues_model.GetLabelsByRepoID(ctx, p.RepoID, "", db.ListOptions{})
```
48e144d015b27c1f8c407bc2964931d0db356899

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+	if (p.Type == project_model.TypeOrganization || p.Type == project_model.TypeIndividual) && p.OwnerID != ctx.ContextUser.ID {
+		ctx.NotFound(nil)
+		return
+	}
```
48e144d015b27c1f8c407bc2964931d0db356899

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,624 @@
+					}
+				}
+			}
+		case "remove_labels":
```
3d5241a260

wxiaoguang commented on models/user/user.go:
```diff
@@ -561,8 +561,9 @@ var globalVars = sync.OnceValue(func() *globalVarsStruct {
-			ActionsUserID: NewActionsUser,
+			GhostUserID:            NewGhostUser,
+			ActionsUserID:          NewActionsUser,
+			ProjectWorkflowsUserID: NewProjectWorkflowsUser,
```
You are just keeping adding "breaking changes" to tell end users that "if you have a user named 'project-workflow', you need to rename it before upgrade", and then "if you have a user named 'xxxx' ...", "if you have a user named 'yyyy' ... "


And you just keep forgetting adding the new names to `reservedUsernames`

<img width="2188" height="508" alt="image" src="https://github.com/user-attachments/assets/21532eab-9657-4d87-b072-484a51a8a314" />

lunny replied:
I think we can also remove those reserved names. A bot account should not have a profile and have a `bot` label which is different from a normal account. i.e. This is an bot account from Github, it will be point to a `/apps/xxxx` but not a normal profile. And there is a `bot` label near the name.
<img width="544" height="713" alt="image" src="https://github.com/user-attachments/assets/b3558932-3df0-4bf0-be1b-d77f229d42c9" />

wxiaoguang replied:
Maybe "at the moment"

OK, that's your choice to keep introducing fragile designs and confusing end users. If you believe it's right, I won't block.

lunny commented on services/projects/workflow_notifier.go:
```diff
@@ -0,0 +1,417 @@
+				log.Error("GetLabelByID: %v", err)
+				continue
+			}
+			toAddedLables = append(toAddedLables, label)
```
They will be checked before saved to database ce3755483d

lunny commented on services/issue/label.go:
```diff
@@ -47,21 +46,6 @@ func AddLabels(ctx context.Context, issue *issues_model.Issue, doer *user_model.
 // RemoveLabel removes a label from issue by given ID.
 func RemoveLabel(ctx context.Context, issue *issues_model.Issue, doer *user_model.User, label *issues_model.Label) error {
 	if err := db.WithTx(ctx, func(ctx context.Context) error {
-		if err := issue.LoadRepo(ctx); err != nil {
```
The permission has been checked outside of this function. So that the duplicated permission check could be removed.

wxiaoguang commented on models/issues/comment.go:
```diff
@@ -236,17 +236,43 @@ func (r RoleInRepo) LocaleHelper(lang translation.Locale) string {
 
+type commentMetaDataKey struct{}
+
+func WithProjectWorkflowCommentMeta(ctx context.Context, meta ProjectWorkflowCommentMeta) context.Context {
```
Why a new context? Why not use `CreateCommentOptions`?

wxiaoguang commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1435 @@
+  margin: 0.125rem;
+}
+
+.ui.labels {
```
Why `.ui.labels`? Is it from Fomantic UI or a pollution to Fomantic UI?

lunny replied:
7c0ac9173b

lunny commented on templates/repo/projects/workflows.tmpl:
```diff
@@ -0,0 +1,12 @@
+	{{template "repo/header" .}}
+	<div class="ui container padded">
+		<div class="tw-flex tw-justify-between tw-items-center tw-mb-4">
+			<a class="ui" href="{{.ProjectLink}}">{{svg "octicon-arrow-left"}} {{ctx.Locale.Tr "projects.workflows"}} {{.Project.Title}}</a>
```
7c0ac9173b

lunny commented on models/issues/comment.go:
```diff
@@ -775,8 +801,17 @@ func (c *Comment) MetaSpecialDoerTr(locale translation.Locale) template.HTML {
 		return locale.Tr("repo.issues.review.codeowners_rules")
+	case SpecialDoerNameProjectWorkflow:
+		if c.CommentMetaData.ProjectWorkflowID > 0 {
+			return htmlutil.HTMLFormat("%s", locale.Tr("repo.issues.project_workflow_action",
```
7c0ac9173b

lunny commented on templates/repo/issue/view_content/comments.tmpl:
```diff
@@ -173,7 +173,7 @@
 					{{template "shared/user/avatarlink" dict "user" .Poster}}
 					<span class="comment-text-line">
-						{{template "shared/user/authorlink" .Poster}}
+						{{template "repo/issue/view_content/comments_authorlink" dict "ctxData" $ "comment" .}}
```
7c0ac9173b

lunny commented on models/migrations/v1_26/v326.go:
```diff
@@ -0,0 +1,25 @@
+
+func AddProjectWorkflow(x *xorm.Engine) error {
+	type ProjectWorkflow struct {
+		ID              int64
```
They are the same from xorm's perspective.

wxiaoguang commented on tests/integration/issue_test.go:
```diff
@@ -122,16 +122,33 @@ func TestNoLoginViewIssue(t *testing.T) {
+		if i < len(opts.LabelIDs)-1 {
+			labelIDsBuf.WriteRune(',')
+		}
+	}
```
`Int64sToStrings`

lunny replied:
ba326099b30c4fe429ea8ff0fd48ac272cfcd583

wxiaoguang replied:
> [ba32609](https://github.com/go-gitea/gitea/commit/ba326099b30c4fe429ea8ff0fd48ac272cfcd583)

Why not just use Int64sToStrings?

wxiaoguang commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,596 @@
+		ctx.HTML(200, tmplRepoWorkflows)
+	} else {
+		ctx.HTML(200, tmplOrgWorkflows)
+	}
```
Why ?

[refactor] replace int with httpStatusCodes (#15282)

lunny replied:
0cc06159269a6669d57ebcb9c487e5ca36aa64cf

wxiaoguang commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,596 @@
+
+	ctx.JSON(http.StatusOK, map[string]any{
+		"success": true,
+	})
```
Who needs this "success"?

Frontend already checks `resp.ok`.

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,596 @@
+	Actions       []project_model.WorkflowAction          `json:"actions"`
+	Summary       string                                  `json:"summary"` // Human readable filter description
+	Enabled       bool                                    `json:"enabled"`
+	IsConfigured  bool                                    `json:"is_configured"` // Whether this workflow is configured/saved
```
As I know, the `typescript` have a habit use `isConfigured` as field?

lunny replied:
I believe the code is consistent, as each layer follows its own naming convention.

In Go, we use `IsConfigured`.
In JSON, we use `is_configured`.
In TypeScript, we use `isConfigured`.

Each naming style is appropriate for its respective context. I don’t think it’s practical to unify all three into a single naming format. Different ecosystems have different conventions and linting rules, and it’s better to follow the standards of each environment rather than force a single style across all layers.

lunny commented on web_src/js/components/projects/WorkflowStore.ts:
```diff
@@ -0,0 +1,430 @@
+        }
+
+        const data = await response.json();
+        const result = camelcaseKeys(data, {deep: true});
```
This example should be a naming bug.

wxiaoguang replied:
I think I have been clear enough: USE the same name between frontend and backend. DO NOT use your fragile toolkit to convert the names.

lunny replied:
3abcb15efb6408433101303c0da77d609b0481fc

lunny commented on routers/web/web.go:
```diff
@@ -1033,7 +1034,14 @@ func registerWebRoutes(m *web.Router) {
+				m.Post("/{workflow_id}", projects.WorkflowsPost)
+				m.Post("/{workflow_id}/status", projects.WorkflowsStatus)
+				m.Post("/{workflow_id}/delete", projects.WorkflowsDelete)
+			}, reqUnitAccess(unit.TypeProjects, perm.AccessModeWrite, true))
```
0cc06159269a6669d57ebcb9c487e5ca36aa64cf

delvh commented on models/issues/comment.go:
```diff
@@ -236,17 +236,44 @@ func (r RoleInRepo) LocaleHelper(lang translation.Locale) string {
-const SpecialDoerNameCodeOwners SpecialDoerNameType = "CODEOWNERS"
+const (
+	SpecialDoerNameCodeOwners      SpecialDoerNameType = "CODEOWNERS"
+	SpecialDoerNameProjectWorkflow SpecialDoerNameType = "ProjectWorkflow"
```
Maybe
```suggestion
	SpecialDoerNameProjectWorkflow SpecialDoerNameType = "automated-action"
```
instead?
After all, it will appear in the timeline, and I don't think `ProjectWorkflow` is the most descriptive name for users.
Alternatively, we could even think about translating this name on demand (not sure if architecturally possible yet)

wxiaoguang replied:
No, it won't appear in the timeline

It is only an internal name.

wxiaoguang replied:
No, none of them would be displayed directly

delvh commented on web_src/js/components/projects/WorkflowStore.ts:
```diff
@@ -0,0 +1,425 @@
+  return result;
+}
+
+function convertFilters(workflow: any): WorkflowFilters {
```
Why are we allowing `any` here, but effectively require it to have a `filters` property?
Wouldn't it make more sense to require a specific type that has the `filters` property?

delvh commented on web_src/js/components/projects/WorkflowStore.ts:
```diff
@@ -0,0 +1,425 @@
+      try {
+        // Load project columns and labels for the dropdowns
+        await store.loadProjectColumns();
+        await store.loadProjectLabels();
```
Also, isn't this a UI route? So, couldn't we combine both calls into one returning both?
Or is there a reason why we are querying them separately?

delvh commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+    store.workflowEvents.splice(tempIndex, 1);
+  }
+
+  if (typeof store.clearDraft === 'function') {
```
Erm… What?
What else could it be?

delvh commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+    window.history.pushState({eventId: event.eventId}, '', newUrl);
+  } catch (error) {
+    console.error('Error selecting workflow event:', error);
+    // On error, try to select the first available workflow instead of clearing
```
Why not no-op?

lunny replied:
updated

delvh commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+  // Find the index of this workflow among same-type workflows
+  const sameTypeIndex = sameTypeWorkflows.findIndex((w: WorkflowListItem) => w.eventId === item.eventId);
+
+  // Extract base name without filter summary (remove anything in parentheses)
```
This is by the way the second time I've seen the ``strip out `(…)` `` code.
Why are we stripping descriptions?
Why not just supply a separate attribute to use?
That way, we don't need to do this hack (twice)

lunny commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+});
+
+// Add debounce mechanism
+let selectTimeout: ReturnType<typeof setTimeout> | null = null;
```
d1ac568

wxiaoguang commented on models/issues/comment.go:
```diff
@@ -802,6 +838,32 @@ func (c *Comment) TimelineRequestedReviewTr(locale translation.Locale, createdSt
+		makeCommentMetaData()
+		commentMetaData.SpecialDoerName = opts.SpecialDoerName
+	}
+	if extDoer, ok := opts.Doer.ExtDoerData.(*projectWorkflowDoer); ok {
```
I believe the tests are not enough.

I ever used `(projectWorkflowDoer)` (intentionally to try) the tests still passes.

lunny commented on tests/integration/issue_test.go:
```diff
@@ -122,16 +123,30 @@ func TestNoLoginViewIssue(t *testing.T) {
+
+	labelIDs := container.FilterSlice(opts.LabelIDs, func(id int64) (string, bool) {
+		return strconv.FormatInt(id, 10), id != 0
+	})
```
f13ae8a2e7

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,558 @@
+	}
+
+	workflowID, _ := strconv.ParseInt(ctx.PathParam("workflow_id"), 10, 64)
+	wf, err := project_model.GetWorkflowByID(ctx, workflowID)
```
f13ae8a2e7

lunny commented on routers/web/projects/workflows.go:
```diff
@@ -0,0 +1,558 @@
+		return
+	}
+
+	workflowID, _ := strconv.ParseInt(ctx.PathParam("workflow_id"), 10, 64)
```
f13ae8a2e7

delvh commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+  await store.loadProjectColumns();
+  await store.loadProjectLabels();
+
+  // Add native event listener to prevent conflicts with Gitea
```
What conflicts are you talking about?
At least give an example in the comment what you're trying to prevent

delvh reviewed (commented):
Here's my review after only looking through the frontend part

delvh commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.event.item_reopened": "Item reopened",
+  "projects.workflows.event.item_closed": "Item closed",
+  "projects.workflows.event.item_column_changed": "Item column changed",
+  "projects.workflows.event.code_changes_requested": "Code changes requested",
```
```suggestion
  "projects.workflows.event.code_changes_requested": "Changes requested",
```

lunny replied:
Why not `code changes requested`?  Even Github used the name.

<img width="249" height="100" alt="image" src="https://github.com/user-attachments/assets/005d035a-9990-4d13-9818-ba83f91c8b96" />

delvh commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.event.item_closed": "Item closed",
+  "projects.workflows.event.item_column_changed": "Item column changed",
+  "projects.workflows.event.code_changes_requested": "Code changes requested",
+  "projects.workflows.event.code_review_approved": "Code review approved",
```
```suggestion
  "projects.workflows.event.code_review_approved": "Approved",
```

delvh commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.apply_to": "Apply to",
+  "projects.workflows.when_moved_from_column": "When moved from column",
+  "projects.workflows.when_moved_to_column": "When moved to column",
+  "projects.workflows.only_if_has_labels": "Only if has labels",
```
Again, shouldn't this string get a parameter instead of concatenation?

lunny replied:
It's the drop down title. it will not have any parameter.

wxiaoguang commented on web_src/js/components/projects/WorkflowStore.ts:
```diff
@@ -0,0 +1,425 @@
+const createDefaultFilters = (): WorkflowFilters => ({issueType: '', sourceColumn: '', targetColumn: '', labels: []});
+const createDefaultActions = (): WorkflowActions => ({column: '', addLabels: [], removeLabels: [], issueState: ''});
+
+const camelToSnake = (key: string): string => key.replace(/([A-Z])/g, '_$1').toLowerCase();
```
The design is simply wrong. 

* https://github.com/go-gitea/gitea/pull/30205#discussion_r2838992987


>I think I have been clear enough: USE the same name between frontend and backend. DO NOT use your fragile toolkit to convert the names.

wxiaoguang commented on web_src/js/components/projects/WorkflowStore.ts:
```diff
@@ -0,0 +1,425 @@
+    async loadEvents(): Promise<WorkflowEvent[]> {
+      const response = await GET(`${props.projectLink}/workflows/events`);
+      const data = await response.json();
+      store.workflowEvents = camelcaseKeys(data, {deep: true}) as WorkflowEvent[];
```
The design is simply wrong. 

* https://github.com/go-gitea/gitea/pull/30205#discussion_r2838992987


>I think I have been clear enough: USE the same name between frontend and backend. DO NOT use your fragile toolkit to convert the names.

lunny replied:
3abcb15efb6408433101303c0da77d609b0481fc

delvh commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.update_workflow_failed": "Failed to update workflow status",
+  "projects.workflows.delete_workflow_failed": "Failed to delete workflow",
+  "projects.workflows.at_least_one_action_required": "At least one action must be configured",
+  "projects.workflows.error.at_least_one_action": "At least one action must be configured",
```
What's the difference between these two?

delvh commented:
Regarding the screenshots you showed:
1. I think the list of workflows on the left should be more compact
2. I'd prefer icons next to the text so that users can quickly identify what event this is without having to read the text (if that is possible which I'm not sure about)

delvh reviewed (commented):
And here are my thoughts regarding the translated strings

lunny commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.filters": "Filters",
+  "projects.workflows.apply_to": "Apply to",
+  "projects.workflows.when_moved_from_column": "When moved from column",
+  "projects.workflows.when_moved_to_column": "When moved to column",
```
These are the drop down titles. They will not have any parameter.

lunny commented on options/locale/locale_en-US.json:
```diff
@@ -3739,6 +3740,55 @@
+  "projects.workflows.close_issue": "Close issue",
+  "projects.workflows.reopen_issue": "Reopen issue",
+  "projects.workflows.save_workflow_failed": "Failed to save workflow",
+  "projects.workflows.update_workflow_failed": "Failed to update workflow status",
```
c90feb817f

lunny commented on web_src/js/components/projects/ProjectWorkflow.vue:
```diff
@@ -0,0 +1,1424 @@
+
+const getCurrentDraftKey = () => {
+  if (!store.selectedWorkflow) return null;
+  return store.selectedWorkflow.eventId || store.selectedWorkflow.workflowEvent;
```
updated

lunny commented:
This is ready to review again.

CreaxxOG commented:
What is the status of this? Would also like to see this merged.

bircni commented:
## Code Review by _Claude Opus 4.7 xHigh_


### High-priority concerns

**1. Workflow actions can re-trigger workflows — cascade / infinite-loop risk**

`executeWorkflowActions` performs state changes (`CloseIssue`, `ReopenIssue`, `MoveIssueToAnotherColumn`, label changes) which themselves emit notifications consumed by `workflowNotifier`. There is **no guard** against the resulting feedback loop. The `feed/notifier.go` change explicitly skips virtual doers (`if doer.ExtDoerData != nil { return }`), but `workflowNotifier` does **not** do the same.

Concrete failure case:
- Workflow A: `on item_closed → reopen issue`
- Workflow B: `on item_reopened → close issue`

`CloseIssue` → `IssueChangeStatus` → fires A → `ReopenIssue` → `IssueChangeStatus` → fires B → ... unbounded recursion.

Column moves happen to be self-limited (`MoveIssueToAnotherColumn` early-returns when `oldColumnID == newColumn.ID`), but close/reopen has no such backstop. Recommend either skipping workflow execution when `doer.ExtDoerData != nil` (consistent with the feed guard), or adding a recursion depth / per-issue dedupe guard.

**2. `RemoveLabel` lost its permission check (`services/issue/label.go`)**

The diff deletes the entire `LoadRepo` + `GetDoerRepoPermission` + `CanWriteIssuesOrPulls` block from the existing `RemoveLabel`. This is a shared function with non-workflow callers:
- `routers/api/v1/repo/issue_label.go:191`
- `routers/web/repo/issue_label.go:218`

Those callers are presumably already behind writer middleware, so this is likely *functionally* safe — but it removes defense-in-depth from a path that previously enforced it, and the change is silent. Please call this out explicitly in the PR description and confirm every caller enforces write permission. The new `AddRemoveLabels` intentionally has no check (for the virtual doer); consider routing workflow label changes through a clearly-named internal helper instead of weakening the public `RemoveLabel`.

**3. Event capabilities are not enforced server-side**

`GetWorkflowEventCapabilities` defines which filters/actions each event supports, but neither `WorkflowsPost` (web) nor `CreateProjectWorkflow`/`UpdateProjectWorkflow` (API) validate the submitted filters/actions against the event's capabilities — `convertFormToFilters` / `convertAPIProjectWorkflowFilters` accept any key. A client can persist e.g. a `source_column` filter or a `column` action on `code_review_approved`, which the executor will then silently process or ignore. Capabilities are currently advisory (UI-only). Recommend validating against `GetWorkflowEventCapabilities()[event]` on create/update.

---

### Medium / lower-priority

- **`IssueChangeProjects` added-project detection** (`workflow_notifier.go`): `if oldProjectColumnMap[newProject.ID] != 0 { continue }` treats column-ID `0` as "not present". If an assigned project can legitimately have column `0`, `item_added_to_project` will mis-fire on unrelated edits. Worth verifying that an assigned project always has a non-zero column ID, or compare key presence explicitly (`_, ok := oldProjectColumnMap[...]`).

- **`executeWorkflowActions` runs outside a transaction** and continues on per-action error → partial application is possible. Probably acceptable for a notifier, but worth a deliberate decision.

- **`ID = -1` collides with `GhostUserID`.** The PR handles this for the feed and `comments_authorlink.tmpl`, but any other code that resolves a comment's `PosterID` could render "Ghost". Consider auditing comment-poster rendering paths, or document the assumption.

- **Inconsistent request handling:** `WorkflowsPost` manually does `io.ReadAll` + `json.Unmarshal` while the API handlers use `web.Bind`. Prefer `web.Bind` for consistency and built-in validation.

- **`UpdateWorkflow`** only updates `workflow_filters`/`workflow_actions` — it cannot change `WorkflowEvent`. That seems intentional (event is fixed per workflow), but confirm.

### Style / conventions

- Test comments reference **absolute line numbers** (`comment_meta_test.go`: *"the type assertion at buildCreateCommentMetaData line 870"*). These rot immediately and violate AGENTS.md's "explain why, never narrate code" guidance — describe the *invariant*, not the location.
- Several new test files carry verbose multi-line explanatory comments; the project prefers short, same-line comments.
- Workflow summary strings in `GetWorkflowSummary` (`"(Issues only)"`, `"(Source: ...)"`, `"(Labels: ...)"`) are **hardcoded English**, not localized — unlike the rest of the feature which uses `locale.Tr`. The summary is surfaced via the API too.

### Positives

- Good test coverage: unit tests for the model, service, capabilities, permission helper (`canWriteProjectWorkflows`), plus integration + e2e tests.
- Clean read/write route separation (`reqRepoReader` vs `reqRepoWriter` / `reqRepoProjectsWriter`).
- `getAPIProjectWorkflowProject` correctly validates the project belongs to the repo (prevents cross-repo access).
- Swagger definitions and migration (#335) are consistent and complete.
- The `feed/notifier.go` virtual-doer guard is a thoughtful fix — just apply the same reasoning to `workflowNotifier` (concern #1).

**Recommendation:** Address #1 (loop guard) and #2 (document/contain the permission-check removal) before merge; #3 is strongly advised. The rest can be follow-ups.

lunny commented:
> ## Code Review by _Claude Opus 4.7 xHigh_
> ### High-priority concerns
> **1. Workflow actions can re-trigger workflows — cascade / infinite-loop risk**
> 
> `executeWorkflowActions` performs state changes (`CloseIssue`, `ReopenIssue`, `MoveIssueToAnotherColumn`, label changes) which themselves emit notifications consumed by `workflowNotifier`. There is **no guard** against the resulting feedback loop. The `feed/notifier.go` change explicitly skips virtual doers (`if doer.ExtDoerData != nil { return }`), but `workflowNotifier` does **not** do the same.
> 
> Concrete failure case:
> 
> * Workflow A: `on item_closed → reopen issue`
> * Workflow B: `on item_reopened → close issue`
> 
> `CloseIssue` → `IssueChangeStatus` → fires A → `ReopenIssue` → `IssueChangeStatus` → fires B → ... unbounded recursion.
> 
> Column moves happen to be self-limited (`MoveIssueToAnotherColumn` early-returns when `oldColumnID == newColumn.ID`), but close/reopen has no such backstop. Recommend either skipping workflow execution when `doer.ExtDoerData != nil` (consistent with the feed guard), or adding a recursion depth / per-issue dedupe guard.
> 
> **2. `RemoveLabel` lost its permission check (`services/issue/label.go`)**
> 
> The diff deletes the entire `LoadRepo` + `GetDoerRepoPermission` + `CanWriteIssuesOrPulls` block from the existing `RemoveLabel`. This is a shared function with non-workflow callers:
> 
> * `routers/api/v1/repo/issue_label.go:191`
> * `routers/web/repo/issue_label.go:218`
> 
> Those callers are presumably already behind writer middleware, so this is likely _functionally_ safe — but it removes defense-in-depth from a path that previously enforced it, and the change is silent. Please call this out explicitly in the PR description and confirm every caller enforces write permission. The new `AddRemoveLabels` intentionally has no check (for the virtual doer); consider routing workflow label changes through a clearly-named internal helper instead of weakening the public `RemoveLabel`.
> 
> **3. Event capabilities are not enforced server-side**
> 
> `GetWorkflowEventCapabilities` defines which filters/actions each event supports, but neither `WorkflowsPost` (web) nor `CreateProjectWorkflow`/`UpdateProjectWorkflow` (API) validate the submitted filters/actions against the event's capabilities — `convertFormToFilters` / `convertAPIProjectWorkflowFilters` accept any key. A client can persist e.g. a `source_column` filter or a `column` action on `code_review_approved`, which the executor will then silently process or ignore. Capabilities are currently advisory (UI-only). Recommend validating against `GetWorkflowEventCapabilities()[event]` on create/update.
> 
> ### Medium / lower-priority
> * **`IssueChangeProjects` added-project detection** (`workflow_notifier.go`): `if oldProjectColumnMap[newProject.ID] != 0 { continue }` treats column-ID `0` as "not present". If an assigned project can legitimately have column `0`, `item_added_to_project` will mis-fire on unrelated edits. Worth verifying that an assigned project always has a non-zero column ID, or compare key presence explicitly (`_, ok := oldProjectColumnMap[...]`).
> * **`executeWorkflowActions` runs outside a transaction** and continues on per-action error → partial application is possible. Probably acceptable for a notifier, but worth a deliberate decision.
> * **`ID = -1` collides with `GhostUserID`.** The PR handles this for the feed and `comments_authorlink.tmpl`, but any other code that resolves a comment's `PosterID` could render "Ghost". Consider auditing comment-poster rendering paths, or document the assumption.
> * **Inconsistent request handling:** `WorkflowsPost` manually does `io.ReadAll` + `json.Unmarshal` while the API handlers use `web.Bind`. Prefer `web.Bind` for consistency and built-in validation.
> * **`UpdateWorkflow`** only updates `workflow_filters`/`workflow_actions` — it cannot change `WorkflowEvent`. That seems intentional (event is fixed per workflow), but confirm.
> 
> ### Style / conventions
> * Test comments reference **absolute line numbers** (`comment_meta_test.go`: _"the type assertion at buildCreateCommentMetaData line 870"_). These rot immediately and violate AGENTS.md's "explain why, never narrate code" guidance — describe the _invariant_, not the location.
> * Several new test files carry verbose multi-line explanatory comments; the project prefers short, same-line comments.
> * Workflow summary strings in `GetWorkflowSummary` (`"(Issues only)"`, `"(Source: ...)"`, `"(Labels: ...)"`) are **hardcoded English**, not localized — unlike the rest of the feature which uses `locale.Tr`. The summary is surfaced via the API too.
> 
> ### Positives
> * Good test coverage: unit tests for the model, service, capabilities, permission helper (`canWriteProjectWorkflows`), plus integration + e2e tests.
> * Clean read/write route separation (`reqRepoReader` vs `reqRepoWriter` / `reqRepoProjectsWriter`).
> * `getAPIProjectWorkflowProject` correctly validates the project belongs to the repo (prevents cross-repo access).
> * Swagger definitions and migration ([Fixed project name on setting module key APP_NAME #335](https://github.com/go-gitea/gitea/pull/335)) are consistent and complete.
> * The `feed/notifier.go` virtual-doer guard is a thoughtful fix — just apply the same reasoning to `workflowNotifier` (concern [Rename import paths: "github.com/gogits/gogs" -> "github.com/go-gitea/gitea" #1](https://github.com/go-gitea/gitea/pull/1)).
> 
> **Recommendation:** Address #1 (loop guard) and #2 (document/contain the permission-check removal) before merge; #3 is strongly advised. The rest can be follow-ups.

All addressed.

lafriks commented:
@lunny would be nice to finish this

lunny committed:
feat(project): rebase project workflow onto main

Recreate `lunny/project_workflow` on top of `main` while dropping the unrelated SSH migration changes that had been mixed into the branch.

lunny commented:
> @lunny would be nice to finish this

Done. I have to force push since the original push broken.

lunny committed:
fix lint

lafriks commented:
If translating human readable description is difficult maybe it could be just user entered description

lunny committed:
Merge branch 'main' into lunny/project_workflow

lunny committed:
update

lunny committed:
update

lunny committed:
Merge branch 'main' into lunny/project_workflow

bircni committed:
cleanup

bircni commented:
did some cleanups

bircni committed:
Merge remote-tracking branch 'origin/main' into lunny/project_workflow

# Conflicts:
#	modelmigration/migrations.go
#	services/projects/issue.go

lunny committed:
fix(project-workflow): bind project lookup to the request route

prepareProject only compared the project owner with ctx.ContextUser, but on a
repo route RepoAssignment sets ContextUser to the repo owner, so owner-level
projects were reachable through /{owner}/{repo}/projects/{id}/workflows.

The owner-scoped write routes were guarded only by reqUnitAccess, which is a
no-op for individual owners, so any signed-in user could modify another user's
personal project workflows. Apply the same individual-owner guard the
neighbouring owner project write group already uses.

lunny committed:
fix(api): populate label exclusive_order in conversions

convert.ToLabel never copied ExclusiveOrder, so every existing label response
reported a wrong value once the field was added to the public struct.

lunny committed:
fix(project): delete project workflows together with their project

DeleteProjectByID and DeleteProjectByRepoID removed columns and issues but left
project_workflow rows behind with no code path able to reach them.

lunny committed:
fix(project-workflow): give the workflow doer a name

Comments and the activity feed special-case the doer, but every other notifier
received a user with an empty name, so webhooks emitted an empty sender login
and the mailer built an empty From display name.

lunny committed:
fix(project-workflow): fire column-changed workflows from the default column

project_issue.project_board_id is 0 for items in the default column, so
GetColumn(ctx, 0) failed and the notifier returned early, silently skipping
item_column_changed workflows for every card dragged out of the default column.

lunny committed:
fix(api): validate project workflow references and keep PATCH partial

Unresolvable column and label references were silently dropped, which widened a
workflow to match everything instead of filtering; they now return 422 like an
unknown issue_type already did.

Filters and actions are pointers now so a PATCH that omits one group no longer
clears it, and the write routes require a non-archived repository like the
equivalent web routes do.

lunny committed:
test(project-workflow): fix imports that broke the test builds

services/projects failed to compile on a missing require import, and the
integration package imported the removed modules/gitrepo package, which blocked
every integration test.

lunny committed:
fix(project-workflow): drop throttle-debounce and fix editor state

throttle-debounce is no longer a dependency, so the bundle and vue-tsc both
failed; use the first-party debounce helper instead.

Cancelling an edit kept the modified draft, so the view showed values the server
never received, and cloning two workflows of the same event produced duplicate
event ids that broke selection, drafts and removal.

lunny committed:
fix(project-workflow): ignore no-op moves inside the default column

Resolving the default column turned oldColumnID 0 into a real id, but the
equality guard had already run against the raw value, so reordering a card
inside the default column looked like a column change.

lunny committed:
refactor(routers): share the individual owner write guard

The check now exists on two project route groups and must not drift.

lunny committed:
fix(project-workflow): select the clicked clone in the sidebar

Unconfigured items were resolved by event type, so clicking the second pending
clone of an event selected the first one.

lunny committed:
fix(user): reserve the project workflow doer name

The name is now rendered as a sender and links to /{username}, so a real
account under it would be indistinguishable from the virtual actor.

lunny committed:
feat(project-workflow): version the stored workflow rules

Filters and actions are JSON blobs with no way to tell a future shape from the
current one, which would force an offline rewrite of every row. Add the
discriminator now, while the table is still empty everywhere, following the
PayloadVersion precedent in models/webhook/hooktask.go.

FindWorkflowsByProjectID had no ORDER BY, so execution order was engine
dependent and on Postgres an UPDATE could silently reorder it.

lunny committed:
fix(project-workflow): give the workflow doer its own system user

The doer used ID -1, which is GhostUserID, so GetPossibleUserFromMap resolved
every workflow comment to the ghost user and generated comments were
indistinguishable from a deleted user's.

lunny committed:
fix(project-workflow): fail closed when a filter cannot be evaluated

A column filter compared with a zero runtime id skipped itself, turning a
restrictive filter into match-all, and a filter type the event never supplies
did the same. Both now refuse to match, and stored filters are checked against
the event capabilities before matching.

Labels were resolved globally by id at execution time, so a workflow kept
applying a label of the previous owner after a repository transfer.

lunny committed:
fix(project-workflow): reject unresolvable references on the web path

Dropping an unresolvable column or label silently removed the filter, so the
workflow matched everything the user meant to narrow, and the request still
reported success. The API already rejected these.

Filters and actions were built by ranging a map, so the stored order differed
between requests, and an unparseable enabled value silently disabled a
workflow.

lunny committed:
fix(api): return project workflow labels like every other endpoint

Both options endpoints hand-built api.Label and shipped the colour with the
leading # that convert.ToLabel strips everywhere else, so clients sharing label
parsing across endpoints broke on this one. Also drop the column colour field,
which no caller reads.

lunny committed:
refactor(project-workflow): split the workflow page into composables

ProjectWorkflow.vue held 427 lines of script behind a 24 line template and
thirteen responsibilities, none of it reachable from a unit test, which is where
the last few bugs lived. Identity, listing, routing, selection and cloning move
into their own modules with vitest coverage.

selectedItem is now derived from the loaded workflow instead of being kept in
sync by hand: seeding it from the deep link made the sidebar report a selection
while the editor pane stayed empty, and the type now makes writing to it a
compile error.

A pending clone no longer survives in the URL, where reloading selected a
different workflow, and no longer disappears when an unrelated workflow is
saved. Label colours are normalised locally now that the server strips the
leading #.

lunny committed:
test(project-workflow): drive the e2e suite through semantic selectors

The suite matched CSS classes exclusively and positioned on the first sidebar
row, and createProject drove two page loads per test.

Two tests whose logic is now covered by vitest are gone; two are new: two
pending clones of one event type, and a reader seeing the page read-only.

lunny committed:
test(project-workflow): group the integration tests under four parents

Sixteen top-level tests each stood up their own environment and rebuilt the same
fixture shape. Consolidating them removes twelve environment reloads.

Three latent couplings had to be resolved first: two children created labels of
the same name on the fixture repository, one derived its issue from whichever
was newest, and the two review children shared a pull request. Verified with
-test.shuffle and by running each child in isolation.

lunny committed:
build: lint the nested vue components

The glob is quoted so stylelint expands it: /bin/sh has no globstar and reduced
it to a single directory level, which dropped files while still exiting 0.

lunny committed:
fix(project-workflow): attribute workflow actions to the triggering user

An earlier commit gave the workflow doer its own system user id. That was the
wrong direction: Gitea has two system users, and CODEOWNERS already shows how to
represent an action that no user performed, by keeping a real user in poster_id
and tagging the comment metadata for display. A synthetic poster_id instead
spreads through every consumer of it, one patch at a time.

The doer is now a copy of the user whose action triggered the workflow, tagged
with the in-memory ExtDoerData that carries the workflow context and drives the
recursion guard. Events without a doer of their own fall back to the issue
poster, as CODEOWNERS does.

The activity feed no longer suppresses these entries: with a real user
attributed there is no reason to hide what a commit-message-triggered close
already shows.

silverwind commented:
One question: This carries "project" in the name but seems only semi-related to gitea projects (kanban). I think we need a better name.

lunny commented:
> One question: This carries "project" in the name but seems only semi-related to gitea projects (kanban). I think we need a better name.

The original name is Project Workflows

silverwind commented:
I would prefer to remove the "project" word, it's the source of confusion. Find another single word.

lunny commented:
> I would prefer to remove the "project" word, it's the source of confusion. Find another single word.

It's nertual as `Project Workflow` because it's the workflow for the projects related behaviours.

silverwind commented:
> workflow for the projects related behaviours

I still don't get it. It seems like a standalone workflow engine. Can this work without the projects unit enabled in a repo? If yes then it should not carry the "projects" name.

Also I don't really see the selling point of this when actions with `on:` triggers exist. Having this stuff committed in a repo is much better. Why do we have 2 ways to define such things?

lunny commented:
> > workflow for the projects related behaviours
> 
> I still don't get it. It seems like a standalone workflow engine. Can this work without the projects unit enabled in a repo? If yes then it should not carry the "projects" name.
> 
> Also I don't really see the selling point of this when actions with `on:` triggers exist. Having this stuff committed in a repo is much better. Why do we have 2 ways to define such things?

No. It can't work without project. The workflow will trigger the project workflow and do some actions when i.e. an issue was assigned to a project.

silverwind commented:
I see, this is coming from https://github.com/go-gitea/gitea/issues/14359. I don't use projects at all in gitea so I'm not the target audience here, but if it depends on it, the name is probably fine.
